Dispatcher Flush Re-Fetch: issues/2478

atopper opened the pull request:
Adding a synchronization type to allow re-fetches on a flush.
- [X] Tested on 6.5.
- [X] Verified installation and smoke tested on 6.3 and 6.4
- [X] Tests added with 91% code coverage
- [X] Updated CHANGELOG

![image](https://user-images.githubusercontent.com/7156029/100385602-17ed8d00-2ff1-11eb-9cd4-da2cea4a3bd1.png)
![image](https://user-images.githubusercontent.com/7156029/100391745-6e16fc00-3002-11eb-97b2-181523b67d43.png)


cc @davidjgonzalez

Issue: https://github.com/Adobe-Consulting-Services/acs-aem-commons/issues/2478

atopper commented on bundle/src/main/java/com/adobe/acs/commons/replication/dispatcher/DispatcherFlushFilter.java:
```diff
@@ -130,13 +130,13 @@ private boolean isEnabled(final Agent agent) {
      */
     private boolean isFlushingAgent(final Agent agent) {
-        return StringUtils.equals(SERIALIZATION_TYPE, agent.getConfiguration().getSerializationType());
+        return StringUtils.startsWith(agent.getConfiguration().getSerializationType(), SERIALIZATION_TYPE);
```
@davidjgonzalez 
I'd like your opinion on this change to existing code.  Instead of having the agent be exactly "flush", the check changed to be starting with "flush".  That'll be backwards compatible.  Will there be collisions?  More than 1 agent trying to do the activation if configured incorrectly?
Thoughts?

atopper commented on bundle/src/main/java/com/adobe/acs/commons/replication/dispatcher/refetchflush/impl/RefetchFlushContentBuilderImpl.java:
```diff
@@ -0,0 +1,421 @@
+
+        String[] uris;
+        int extSep = path.indexOf('.', path.lastIndexOf('/'));
+        if (extSep == -1) {
```
I thought about other additions here, to help make the activation more powerful and flexible for the user, but thought I'd smart small.

davidjgonzalez commented:
@atopper can you take a quick look at: https://github.com/davidjgonzalez/acs-aem-commons/commit/9957874e15519aed4a621627484cb5ed5702d870#
I updated to use new OSGi annotations.

Have you explored using the ReplicationContentFacade instead of a tmp file? I believe that lets you store the data in the JCR instead of a tmp file - though, i guess that would need service user access to the node? We could put them under /tmp so theyre auto-cleaned up?

https://helpx.adobe.com/experience-manager/6-3/sites/developing/using/reference-materials/javadoc/com/day/cq/replication/ReplicationContentFacade.html

atopper commented:
@davidjgonzalez Thanks for the review!  I didn't mean to make you do so much work!
I pushed the new OSGi annotations, and added @Designate and a few more properties to make it look good in the /system UI, and so the ContentBuilder could be found by the Agent.
I removed the AUTH_INFO.

Regarding **ReplicationContentFacade**, I like the idea - not a lot of documentation around that one but I played around with it.  I worry about the service user/ACLs that you mentioned.  I also can't find how often the /tmp folder is cleaned.  Customers could activate a lot of content, so I feel like I should implement folder buckets too.  I see createTempFile used in CQ and Granite.
How strongly do you feel that **ReplicationContentFacade** be used.  It feels like the complexity increases a lot.

davidjgonzalez commented:
Let me take a look at this tomorrow and see if we can't cut a release
